LifecycleEvents prototype

Arkensor opened the pull request:
The addition of `CF_LifecycleEvents` reduces boiler code and avoids merge conflicts for connecting features to lifecycle functions that they need for init or cleanup. Instead of the less optimal central function that has to call all `<feature>._Cleanup()` methods, it is now easily possible to, for example, subscribe to OnMissionEnd via an attribute above the concrete method inside the feature code.

```cs
class LifecycleStaticDemo
{
    [CF_EventSubscriber(LifecycleStaticDemo.GameCreate, CF_LifecycleEvents.OnGameCreate)]
    static void GameCreate()
    {
        Print("LifecycleStaticDemo::GameCreate");
    }

    [CF_EventSubscriber(LifecycleStaticDemo.GameDestroy, CF_LifecycleEvents.OnGameDestroy)]
    static void GameDestroy()
    {
        Print("LifecycleStaticDemo::GameDestroy");
    }

    [CF_EventSubscriber(LifecycleStaticDemo.MissionCreate, CF_LifecycleEvents.OnMissionCreate)]
    static void MissionCreate()
    {
        Print("LifecycleStaticDemo::MissionCreate");
    }

    [CF_EventSubscriber(LifecycleStaticDemo.MissionDestroy, CF_LifecycleEvents.OnMissionDestroy)]
    static void MissionDestroy()
    {
        Print("LifecycleStaticDemo::MissionDestroy");
    }

    [CF_EventSubscriber(LifecycleStaticDemo.OnePrintToRuleThemAll,
        CF_LifecycleEvents.OnGameCreate,
        CF_LifecycleEvents.OnGameDestroy,
        CF_LifecycleEvents.OnMissionCreate,
        CF_LifecycleEvents.OnMissionDestroy)]
    static void OnePrintToRuleThemAll()
    {
        Print("OnePrintToRuleThemAll");
    }

    [CF_MultiEventSubscriber(LifecycleStaticDemo.TestMultiEventSubscriber,
    {
        CF_LifecycleEvents.OnGameCreate,
        CF_LifecycleEvents.OnMissionCreate,
        CF_LifecycleEvents.OnMissionDestroy,
        CF_LifecycleEvents.OnGameDestroy
    })]
    static void TestMultiEventSubscriber()
    {
        Print("TestMultiEventSubscriber");
    }
};
```

```
SCRIPT       : CF_CreateGame()
SCRIPT       : ... Backlit Effects Enabled
SCRIPT       : LifecycleInstanceDemo::HandleGameCreate      <<<<<
SCRIPT       : LifecycleStaticDemo::GameCreate              <<<<<
SCRIPT       : OnePrintToRuleThemAll                        <<<<<
SCRIPT       : TestMultiEventSubscriber                     <<<<<
SCRIPT       : LifecycleStaticDemo::MissionCreate           <<<<<    
SCRIPT       : OnePrintToRuleThemAll                        <<<<<
SCRIPT       : TestMultiEventSubscriber                     <<<<<
SCRIPT       : Creating ModuleManager                       <<<<<
SCRIPT       : CF_CreateWorld()
SCRIPT       : Modules Loaded
SCRIPT       : SyncEvents -> RegisterEvents
SCRIPT       : AnimSoundObjectBuilderBank: Invalid sound set "generalgrunt_SoundVoice_Char_SoundSet".
[...]
SCRIPT       : OnMissionFinish
SCRIPT       : LifecycleStaticDemo::MissionDestroy          <<<<<
SCRIPT       : OnePrintToRuleThemAll                        <<<<<
SCRIPT       : TestMultiEventSubscriber                     <<<<<
SCRIPT       : LifecycleStaticDemo::GameDestroy             <<<<<
SCRIPT       : OnePrintToRuleThemAll                        <<<<<
SCRIPT       : TestMultiEventSubscriber                     <<<<<
SCRIPT       : ~DayZGame()
```

As an underlying construct, a small `ScriptInvoker` wrapper called `CF_EventHandler`was added that is inspired by [C#'s EventHandler concept](https://docs.microsoft.com/en-us/dotnet/api/system.eventhandler-1). It essentially introduces a sender object and strongly typed event arguments. The type of arguments can be defined via the template parameter. Anything non-primitive should be inherited from the `CF_EventArgs` base class.
```cs
void EventArgsDemo()
{
    CF_EventHandlerT<SomeEventArgs> evt();
    evt.AddSubscriber(Handle);
    evt.Invoke(GetGame(), new SomeEventArgs());
}

class SomeEventArgs : CF_EventArgs {}

void Handle(Class sender, SomeEventArgs arguments)
{
    Print(arguments);
}
```
Because of the way `ScriptInvoker` works, a function that takes only the sender or no arguments at all can also subscribe and will be invoked. Similar to how it works in C#.

Arkensor committed:
Added LifecycleEvents prototype

Jacob-Mango reviewed (changes requested):
CF_LifecycleEventSubscription is too long, I suggest renaming to CF_LifecycleEvent

Arkensor committed:
Moved attribute to EventHandler core functionality instead of LifecycleEvents

Arkensor committed:
Added EventHandlerT variant
Updated default EventHandler to use EventHandlerT<CF_EventArgs>

Arkensor committed:
Updated LifeCycleEvents to use non template EventHandler version

Arkensor committed:
Added inital documentation for EventHandler, LifecycleEvents
Moved CF_EventArgs, CF_EventHandler, CF_LifecycleEvents into 2_GameLib module so they are guranteed available in 3_Game no matter the load order
Added protected ctor/dtor in C_LifecycleEvents to avoid people from accidently wanting to instantiate the class
Renamed CF_EventHandler Subscribe() to AddSubscriber()
Added EventHandler default NULL parameters to Invoke()
Renamed CF_EventSubscription to CF_EventSubscriber and added Multi-Array variant of it

Arkensor committed:
Fixed docs link from LifecycleEvents to EventHandler
